tools/aot/execute_test_generator.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+# Also available under a BSD-style license. See LICENSE.
+
+import re
+import argparse
+import numpy as np
+
+parser = argparse.ArgumentParser(
+    description="Test generator for AOT compiled programs using a test template"
+)
+parser.add_argument(
+    "--test_template_path",
+    required=True,
+    help="Path to the execute_test.template.cpp file",
+)
+parser.add_argument(
+    "--reference_tensors_path",
+    required=True,
+    help="Path to the file containing the reference inputs and outputs (.npz)",
+)
+
+numpy_to_memref_dtype_map = {
+    # Add more mappings as needed
+    "int8": "int8_t",
+    "uint8": "uint8_t",
+    "int16": "int16_t",
+    "uint16": "uint16_t",
+    "int32": "int32_t",
+    "uint32": "uint32_t",
+    "int64": "int64_t",
+    "uint64": "uint64_t",
+    "float32": "float",
+    "float64": "double",
+}
+
+
+def main():
+    args = parser.parse_args()
+
+    # Track string substitutions
+    output_memref_variable_declarations_str = ""
+    input_memref_abi_declarations_str = ""
+    read_reference_tensors_into_npy_array_str = ""
+    create_memref_from_npy_array_str = ""
+    pass_input_memref_arguments_str = ""
+    assert_result_shape_matches_reference_str = ""
+    expect_result_data_matches_reference_str = ""
+    deallocate_result_memref_str = ""
+    reference_tensors_path_str = args.reference_tensors_path.removeprefix(
+        "bazel-out/k8-fastbuild/bin/"
+    )
+
+    # Interpret function signature (num_args, rank, dtype, num_returns)
+    # from the saved reference tensors and build string substitutions
+    reference_tensors = np.load(args.reference_tensors_path)
+    for key in reference_tensors.keys():
+        tensor = reference_tensors[key]
+        rank = tensor.ndim
+        dtype = numpy_to_memref_dtype_map[str(tensor.dtype)]
+
+        if "Input" in key:
+            input_memref_abi_declarations_str += f"""
+    DECL_RANK_{rank}_MEMREF_ABI({dtype}),"""
+            pass_input_memref_arguments_str += f"""
+      PASS_RANK_{rank}_MEMREF({key}),"""
+            if rank == 0:
+                create_memref_from_npy_array_str += f"""
+  StridedMemRefType<{dtype}, {rank}> {key} =
+      CreateMemRefFromNpyArray<{dtype}>(ref{key});"""
+            else:
+                create_memref_from_npy_array_str += f"""
+  StridedMemRefType<{dtype}, {rank}> {key} =
+      CreateMemRefFromNpyArray<{dtype}, {rank}>(ref{key});"""
+
+        if "Output" in key:
+            output_memref_variable_declarations_str += f"""
+  StridedMemRefType<{dtype}, {rank}> {key};"""
+            for n in range(rank):
+                assert_result_shape_matches_reference_str += f"""
+  ASSERT_EQ(Result.{key}.sizes[{n}], ref{key}.shape[{n}]);"""
+            expect_result_data_matches_reference_str += f"""
+  for (int i = 0; i < ref{key}.num_vals; i++)
+    EXPECT_EQ(Result.{key}.data[i], ref{key}.data<{dtype}>()[i]);"""
```

Comment:
![image](https://github.com/cruise-automation/mlir-tcp/assets/19234106/d0db8603-b58c-430a-8e51-82c336d70f7d)
https://github.com/cruise-automation/mlir-tcp/commit/ada575353cd2381504447c93bc204773134a9fc2#r139405522

Reply:
@srinathava for the simple case of elementwise add/mul, `EXPECT_EQ` seems to work (example: `actual: 9.97501659 vs 9.97501659`) but I can see this falling apart for anything slightly more complex. I am considering using one of [these](https://google.github.io/googletest/reference/assertions.html#floating-point) - `EXPECT_FLOAT_EQ` or `EXPECT_DOUBLE_EQ` for non-integer typed outputs. Any concerns with that?`EXPECT_NEAR` requires an absolute error bound, which may not be easy to determine consistently for varying dynamic ranged tensors.

Reply:
Done, PTAL (https://github.com/cruise-automation/mlir-tcp/pull/46/commits/9796a7e4d6223c2af6086ca9feabbf44edc15d54).